Shadow demo and Fleet provisioning demo modifations.

zhangzz0413 opened the pull request:
Shadow demo modifations.
1. Modify the original demo to shadow_demo_with_openssl.
2. Add shadow demo used mbedtls and pkcs11.
3. Add shadow demo only used mbedtls.

*Issue #, if available:*

*Description of changes:*


By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

zhangzz0413 committed:
Shadow demo modifations.
1. Modify the original demo to shadow_demo_with_openssl.
2. Add shadow demo used mbedtls and pkcs11.
3. Add shadow demo only used mbedtls.

zhangzz0413 committed:
Fleet provisioning demo modifications.
1. Add fleet provisioning crt demo which called CreateKeysAndCertificate interface.
2. Add fleet provisioning crt demo used mbedtls but not used pkcs11.
3. Add fleet provisioning csr demo used mbedtls but not used pkcs11.

zhangzz0413 committed:
Modified the issue of failed greengrass local auth demo.

zhangzz0413 committed:
Modify mqtt_demo_mutual_auth demo.
1. Add a thread for looping call MQTT_processLoop;
2. Use a mutex to protect the functions of the coreMQTT Library;
3. Add a publish failed resend mechanism, if resends three times fail, to reconnect MQTT broker;

zhangzz0413 committed:
Modify mqtt_demo_mutual_auth demo.
1. Add a thread for looping call MQTT_processLoop;
2. Use a mutex to protect the functions of the coreMQTT Library;
3. Add a publish failed resend mechanism, if resends three times fail, to reconnect MQTT broker;

awsblake commented on demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c:
```diff
@@ -167,7 +173,7 @@
  * @brief Timeout for receiving CONNACK packet in milli seconds.
  */
-#define CONNACK_RECV_TIMEOUT_MS                  ( 1000U )
+#define CONNACK_RECV_TIMEOUT_MS                  ( 2000U )
```
we can restore this change due to network latency.

awsblake commented on demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c:
```diff
@@ -358,6 +373,29 @@ static MQTTPubAckInfo_t pOutgoingPublishRecords[ OUTGOING_PUBLISH_RECORD_LEN ];
+/**
+ * @brief Keep a flag for indicating if the MQTT connection is alive.
+ */
+static bool mqttStartProcessLoop = false;
```
/**
 * @brief Keep a flag for indicating if the MQTT connection is alive.
 */
static bool mqttSessionEstablished = false;

awsblake commented on demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c:
```diff
@@ -366,6 +404,16 @@ struct NetworkContext
 };
 
+/**
+ * @brief MQTT Context for mqtt connection.
```
MQTT connection context used in this demo.

awsblake commented on demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c:
```diff
@@ -366,6 +404,16 @@ struct NetworkContext
+MQTTContext_t mqttContext = { 0 };
+
+/**
+ * @brief Network context for mqtt connection.
```
Network connection context used in this demo.

awsblake commented on demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c:
```diff
@@ -1185,14 +1288,31 @@ static int disconnectMqttSession( MQTTContext_t * pMqttContext )
-                    MQTT_Status_strerror( mqttStatus ) ) );
-        returnStatus = EXIT_FAILURE;
+        /* Clear the mqtt session flag. */
+        mqttStartProcessLoop = false;
```
Shall we set mqttStartProcessLoop to false after MQTT_Disconnect?

awsblake commented on demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c:
```diff
@@ -1185,14 +1288,31 @@ static int disconnectMqttSession( MQTTContext_t * pMqttContext )
-    mqttStatus = MQTT_Disconnect( pMqttContext );
-
-    if( mqttStatus != MQTTSuccess )
+    if ( mqttStartProcessLoop )
```
Write a disconnect function and put all the code into there

awsblake commented on demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c:
```diff
@@ -1456,23 +1612,42 @@ static int subscribePublishLoop( MQTTContext_t * pMqttContext )
+                 * has expired since the last MQTT packet sent and receive
+                 * ping responses. */
+                returnStatus = waitForAck();
+                if ( returnStatus != EXIT_SUCCESS )
```
move this to a new function called mqttRepublish?

awsblake commented on demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c:
```diff
@@ -1494,9 +1669,7 @@ static int subscribePublishLoop( MQTTContext_t * pMqttContext )
-        returnStatus = waitForPacketAck( pMqttContext,
-                                         globalUnsubscribePacketIdentifier,
-                                         MQTT_PROCESS_LOOP_TIMEOUT_MS );
+        returnStatus = waitForAck();
```
Keep the function name as waitForPacketAck

zhangzz0413 committed:
Modify mqtt_demo_mutual_auth demo.
1. Add a thread for looping call MQTT_processLoop;
2. Use a mutex to protect the functions of the coreMQTT Library;
3. Add a publish failed resend mechanism, if resends three times fail,
to reconnect MQTT broker.

zhangzz0413 committed:
Modify mqtt_demo_mutual_auth demo.
1. Add a thread for looping call MQTT_processLoop;
2. Use a mutex to protect the functions of the coreMQTT Library;
3. Add a publish failed resend mechanism, if resends three times fail,
to reconnect MQTT broker.
4. Modification of review comments.

zhangzz0413 committed:
Revert code for recommit.

zhangzz0413 committed:
Modify mqtt_demo_mutual_auth demo.
1. Add a thread for looping call MQTT_processLoop.
2. Use a mutex to protect the functions of the coreMQTT Library;
3. Add a publish failed resend mechanism, if resends three times fail,
to reconnect MQTT broker.
4. Modification of review comments.
